test/functional/p2p_permissions.py

```diff
@@ -83,6 +97,41 @@ def run_test(self):
         self.nodes[1].assert_start_raises_init_error(["-whitelist=noban@127.0.0.1:230"], "Invalid netmask specified in", match=ErrorMatch.PARTIAL_REGEX)
         self.nodes[1].assert_start_raises_init_error(["-whitebind=noban@127.0.0.1/10"], "Cannot resolve -whitebind address", match=ErrorMatch.PARTIAL_REGEX)
 
+    def check_tx_relay(self):
+        block_op_true = self.nodes[0].getblock(self.nodes[0].generatetoaddress(100, ADDRESS_BCRT1_P2WSH_OP_TRUE)[0])
+        self.sync_all()
+
+        self.log.debug("Create a connection from a whitelisted wallet that rebroadcasts raw txs")
+        # A python mininode is needed to send the raw transaction directly. If a full node was used, it could only
+        # rebroadcast via the inv-getdata mechanism. However, even for whitelisted connections, a full node would
+        # currently not request a txid that is already in the mempool.
+        self.restart_node(1, extra_args=["-whitelist=forcerelay@127.0.0.1"])
+        p2p_rebroadcast_wallet = self.nodes[1].add_p2p_connection(P2PDataStore())
+
+        self.log.debug("Send a tx from the wallet initially")
+        tx = FromHex(
+            CTransaction(),
+            self.nodes[0].createrawtransaction(
+                inputs=[{
+                    'txid': block_op_true['tx'][0],
+                    'vout': 0,
+                }], outputs=[{
+                    ADDRESS_BCRT1_P2WSH_OP_TRUE: 5,
+                }]),
+        )
+        tx.wit.vtxinwit = [CTxInWitness()]
+        tx.wit.vtxinwit[0].scriptWitness.stack = [CScript([OP_TRUE])]
+        txid = tx.rehash()
+
+        self.log.debug("Wait until tx is in node[1]'s mempool")
+        p2p_rebroadcast_wallet.send_txs_and_test([tx], self.nodes[1])
+
+        self.log.debug("Check that node[1] will send the tx to node[0] even though it is already in the mempool")
+        connect_nodes(self.nodes[1], 0)
+        with self.nodes[1].assert_debug_log(["Force relaying tx {} from whitelisted peer=0".format(txid)]):
+            p2p_rebroadcast_wallet.send_txs_and_test([tx], self.nodes[1])
+            wait_until(lambda: txid in self.nodes[0].getrawmempool())
```

Comment:
Is this what we really want? If the tx is already in node 0's mempool, wouldn't this always return right away?

Reply:
Oh, I think my comment is unclear. It refers to node[1]'s mempool.

So node[1] will rebroadcast the tx from the wallet even though it is already in node[1]'s mempool

Reply:
ah! that makes a lot more sense.